lib/upsert/merge_function/postgresql.rb

```diff
@@ -40,13 +40,159 @@ def clear(connection)
+        begin
+          execute_parameterized(sql, values.map { |v| connection.bind_value v })
+        rescue self.class::ERROR_CLASS => pg_error
+          if pg_error.message =~ /function #{name}.* does not exist/i
+            if first_try
+              Upsert.logger.info %{[upsert] Function #{name.inspect} went missing, trying to recreate}
+              first_try = false
+              create!
+              retry
+            else
+              Upsert.logger.info %{[upsert] Failed to create function #{name.inspect} for some reason}
+              raise pg_error
+            end
+          else
+            raise pg_error
+          end
+        end
+      end
+
+      # strangely ? can't be used as a placeholder
       def sql
         @sql ||= begin
-          bind_params = Array.new(selector_keys.length + setter_keys.length, '?') + Array.new(hstore_delete_handlers.length, '?::text[]')
+          bind_params = []
+          i = 1
+          (selector_keys.length + setter_keys.length).times do
+            bind_params << "$#{i}"
+            i += 1
+          end
+          hstore_delete_handlers.length.times do
+            bind_params << "$#{i}::text[]"
+            i += 1
+          end
           %{SELECT #{name}(#{bind_params.join(', ')})}
         end
       end
 
+      def use_pg_native?
+        server_version >= 95 && unique_index_on_selector?
+      end
+
+      def server_version
+        @server_version ||=
+          controller.connection.execute("SHOW server_version").first["server_version"].split('.')[0..1].join('').to_i
+      end
+
+      def schema_query
+        execute_parameterized(
+          %{
+            SELECT array_agg(column_name::text) AS index_columns FROM information_schema.constraint_column_usage
+              JOIN pg_catalog.pg_constraint ON constraint_name::text = conname::text
+              WHERE table_name = $1 AND conrelid = $1::regclass::oid AND contype = 'u'
+              GROUP BY table_catalog, table_name, constraint_name
+          },
+          [table_name]
+        )
+      end
+
+      def pg_native(row)
+        bind_setter_values = row.setter.values.map { |v| connection.bind_value v }
+
+        upsert_sql = %{
+          INSERT INTO #{quoted_table_name} (#{quoted_setter_names.join(',')})
+          VALUES (#{insert_bind_placeholders(row).join(', ')})
+          ON CONFLICT(#{quoted_selector_names.join(', ')})
+          DO UPDATE SET (#{quoted_setter_names.join(', ')}) = (#{conflict_bind_placeholders(row).join(', ')})
+        }
+
+        execute_parameterized(upsert_sql, bind_setter_values)
+      end
+
+      def hstore_delete_function(sql, row, column_definition)
+        parts = []
+        if row.hstore_delete_keys.key?(column_definition.name)
+          parts << "DELETE("
+        end
+        parts << sql
+        if row.hstore_delete_keys.key?(column_definition.name)
+          keys = row.hstore_delete_keys[column_definition.name].map { |k| "'#{k.to_s.gsub("'", "\\'")}'" }
```

Comment:
better not to add a dependency if you are 100% sure you are getting the quoting right and complete